File: test/selenium/swisssearch_test.js
// Swisssearch Test Using Browserstack

var webdriver = require('browserstack-webdriver');
var assert = require('assert');

var QUERYSTRING_OF_RARON = "X=128176.00&Y=629766.00&zoom=10";
var QUERYSTRING_OF_RTE_BERNE_LAUSANNE = "X=154208.00&Y=539257.00&zoom=10";
var QUERYSTRING_OF_PL_CHATEAU_AVENCHES = "X=192310.00&Y=569734.00&zoom=10";
var QUERYSTRING_OF_PIAZZA_MESOLCINA_BELLINZONA = "X=117500.00&Y=722500.00&zoom=10";
var QUERYSTRING_MOOS = "X=128630.00&Y=627650.00&zoom=10";
var QUERYSTRING_OF_REALTA = "X=181085.00&Y=751355.00";

var runTest = function(cap, driver, target){
  //swissearch parameter with multiple results
  driver.get(target + '/?swisssearch=raron&lang=de');
  //wait until topics related stuff is loaded. We know this when catalog is there
  driver.findElement(webdriver.By.xpath("//a[contains(text(), 'Grundlagen und Planung')]"));
  driver.findElement(webdriver.By.xpath("//*[contains(text(), 'Raron')]"));
  driver.findElement(webdriver.By.xpath("//*[contains(text(), ', Flugplatz')]")).click();
  driver.findElement(webdriver.By.xpath("//a[contains(@href, '" + QUERYSTRING_OF_RARON + "')]"));
  //parameter should disappear when selection is done
  driver.findElement(webdriver.By.xpath("//*[@id='toptools']//a[contains(@href,'http')]")).getAttribute("href").then(function(val) {
    assert.ok(val.indexOf('swisssearch') == -1);
  });

  //swisssearch parameter with multiple results (locations and layers), reset selection
  driver.get(target + '/?swisssearch=wasser&lang=de');
  //wait until topics related stuff is loaded. We know this when catalog is there
  driver.findElement(webdriver.By.xpath("//a[contains(text(), 'Grundlagen und Planung')]"));
  driver.findElement(webdriver.By.xpath("//*[contains(text(), 'Gehe nach')]"));
  driver.findElement(webdriver.By.xpath("//*[contains(text(), 'Karte hinzufügen')]"));
  driver.findElement(webdriver.By.xpath("//button[@ng-click='clearInput()']")).click();
  //parameter should disappear when selection is done
  driver.findElement(webdriver.By.xpath("//*[@id='toptools']//a[contains(@href,'http')]")).getAttribute("href").then(function(val) {
    assert.ok(val.indexOf('swisssearch') == -1);
  });

  //swisssearch Realta industriegebiet 701 with wordforms GI -> industriegebiet (DE)
  driver.get(target + '/?swisssearch=realta gi 701&lang=de');
  //wait until topics related stuff is loaded. We know this when catalog is there
  driver.findElement(webdriver.By.xpath("//a[contains(text(), 'Grundlagen und Planung')]"));
  driver.findElement(webdriver.By.xpath("//a[contains(@href, '" + QUERYSTRING_OF_REALTA + "')]"));

  //swisssearch Route de Berne 91 1010 Lausanne with wordforms rte
  driver.get(target + '/?swisssearch=rte berne 91 1010&lang=de');
  //wait until topics related stuff is loaded. We know this when catalog is there
  driver.findElement(webdriver.By.xpath("//a[contains(text(), 'Grundlagen und Planung')]"));
  driver.findElement(webdriver.By.xpath("//a[contains(@href, '" + QUERYSTRING_OF_RTE_BERNE_LAUSANNE + "')]"));

  //swisssearch Place du Château, Avenches (Pl -> platz)
  driver.get(target + '/?swisssearch=pl chateau avenches&lang=de');
  //wait until topics related stuff is loaded. We know this when catalog is there
  driver.findElement(webdriver.By.xpath("//a[contains(text(), 'Grundlagen und Planung')]"));
  driver.findElement(webdriver.By.xpath("//a[contains(@href, '" + QUERYSTRING_OF_PL_CHATEAU_AVENCHES + "')]"));

  //swisssearch Piazza Mesolcina, Bellinzona (P -> piazza)
  driver.get(target + '/?swisssearch=p Mesolcina Bellinzona&lang=de');
  //wait until topics related stuff is loaded. We know this when catalog is there
  driver.findElement(webdriver.By.xpath("//a[contains(text(), 'Grundlagen und Planung')]"));
  driver.findElement(webdriver.By.xpath("//a[contains(@href, '" + QUERYSTRING_OF_PIAZZA_MESOLCINA_BELLINZONA + "')]"));

  //swissearch parameter with 1 result (direct selection doesn't work in safari 5.1)
  driver.get(target + '/?swisssearch=brückenmoostrasse 11 raron&lang=de');
  //wait until topics related stuff is loaded. We know this when catalog is there
  driver.findElement(webdriver.By.xpath("//a[contains(text(), 'Grundlagen und Planung')]"));
  driver.findElement(webdriver.By.xpath("//a[contains(@href, '" + QUERYSTRING_MOOS + "')]"));
  //parameter stays after initial automatic selection
  driver.findElement(webdriver.By.xpath("//*[@id='toptools']//a[contains(@href,'http')]")).getAttribute("href").then(function(val) {
    assert.ok(val.indexOf(QUERYSTRING_MOOS) > -1);
    assert.ok(val.indexOf('swisssearch=br') > -1);
  });
  //parameter is removed by map action (simulating zoom here)
  driver.findElement(webdriver.By.xpath("//button[@class='ol-zoom-in']")).click();
  driver.findElement(webdriver.By.xpath("//a[contains(@href, 'zoom=11')]"));
  //parameter should disappear when selection is done
  driver.findElement(webdriver.By.xpath("//*[@id='toptools']//a[contains(@href,'http')]")).getAttribute("href").then(function(val) {
    assert.ok(val.indexOf('swisssearch') == -1);
  });
}

module.exports.runTest = runTest;

File: test/saucelabs/swisssearch_test.py
# -*- coding: utf-8 -*-

import time
import sys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC

DEFAULT_WAIT = 10
# Gloabl css selectors
inputContainerClass = 'ga-search-input-container'
dropDownClass = 'ga-search-dropdown'
itemClass = 'ga-search-item'


def waitUrlChange(driver, pattern, find=True, timeout=DEFAULT_WAIT):
    t0 = time.time()
    newUrl = driver.current_url
    if find:
        # We wait until we find pattern
        while pattern not in newUrl:
            newUrl = driver.current_url
            t1 = time.time()
            if t1 - t0 > timeout:
                return True
            time.sleep(.5)
        return False
    else:
        # We wait until we don't find a pattern
        while pattern in newUrl:
            newUrl = driver.current_url
            t1 = time.time()
            if t1 - t0 > timeout:
                return True
            time.sleep(.5)
        return False


def onLocationFail(currentUrl, searchText, resultLocation):
    print 'The map has not been zoomed at the expected location.'
    print 'searchText: %s' % searchText
    print 'Expect to find: %s' % resultLocation
    print 'In: %s' % currentUrl
    sys.exit(1)


def selectLocationItem(driver, inputEl, searchText, resultTitle, resultLocation):
    try:
        # We expect to have the dropdown open
        WebDriverWait(driver, 10).until(
            EC.visibility_of_element_located((By.CLASS_NAME, dropDownClass)))
    except Exception as e:
        print 'Could not toggle dropdown visibility using searchText: %s' % searchText
        print 'Input value is: %s' % inputEl.get_attribute('value')
        raise Exception(e)
    searchResultEls = driver.find_elements_by_class_name(itemClass)
    for resultEl in searchResultEls:
        if resultTitle == resultEl.get_attribute('title'):
            resultEl.click()
            break
    # Make sure map is zoomed to the desired location
    if waitUrlChange(driver, resultLocation):
        onLocationFail(driver.current_url, searchText, resultLocation)


def testSearchInput(driver, inputEl, searchTest, timeout=DEFAULT_WAIT):
    searchText = searchTest['searchText']
    resultTitle = searchTest['resultTitle']
    resultLocation = searchTest['resultLocation']
    inputEl.send_keys(searchText)
    # Because of the search design we have to trigger a change event
    # to trigger a search request
    driver.execute_script("$('ga-search-input').change()")
    selectLocationItem(driver, inputEl, searchText, resultTitle, resultLocation)

    # Use clear button
    driver.find_element_by_css_selector(
        '.%s .ga-btn' % inputContainerClass).click()
    try:
        WebDriverWait(driver, 10).until(
            EC.invisibility_of_element_located((By.CLASS_NAME, dropDownClass)))
    except Exception as e:
        print 'Could not hide dropdown visibility using the search button'
        raise Exception(e)
    # Make sure input field is empty
    if inputEl.get_attribute('value') != '':
        print 'Clear button in search input did not work as expected'
        sys.exit(1)

    print 'Search test for %s OK!' % searchText


def testSwissSearchParameter(driver, url, inputEl, searchTest, timeout=DEFAULT_WAIT):
    searchText = searchTest['searchText']
    oneResOnly = searchTest['oneResOnly']
    resultTitle = searchTest['resultTitle']
    resultLocation = searchTest['resultLocation']
    targetUrl = url + '/?lang=de&swisssearch=%s' % searchText
    driver.get(targetUrl)
    # One result only -> zoom to the desired location
    if oneResOnly:
        if waitUrlChange(driver, resultLocation):
            onLocationFail(driver.current_url, searchText, resultLocation)
        try:
            assert "swisssearch" in driver.current_url
        except AssertionError as e:
            raise Exception(e)
        # parameter (swisssearch) is removed by map action (simulating zoom)
        driver.find_element_by_css_selector("button.ol-zoom-out").click()
    else:
        selectLocationItem(driver, inputEl, searchText, resultTitle, resultLocation)
    if waitUrlChange(driver, 'swisssearch', find=False):
        print 'swissearch param can still be found in the URL after a map move'
        sys.exit(1)

    print 'SwissSearch parameter test for %s OK!' % searchText


searchLocationTests = [
    {
        'searchText': u'Kanalstrasse',
        'resultTitle': u'Bus Raron, Kanalstrasse',
        'resultLocation': u'X=128141.00&Y=627443.00&zoom=10'
    },
    {
        'searchText': u'p Mesolcina Bellinzona',
        'resultTitle': u'Bus Bellinzona, Piazza Mesolcina',
        'resultLocation': u'X=117500.00&Y=722500.00&zoom=10'
    },
    {
        'searchText': u'bruckenmoostrasse 11 raron',
        'resultTitle': u'Brückenmoosstrasse 11 3942 Raron',
        'resultLocation': u'X=128630.12&Y=627650.38&zoom=10'
    },
    {
        'searchText': u'rte berne 91 1010',
        'resultTitle': u'Route de Berne 91 1010 Lausanne',
        'resultLocation': u'X=154207.56&Y=539257.38&zoom=10'
    },
    {
        'searchText': u'pl chateau 3 laus',
        'resultTitle': u'Place du Château 3 1005 Lausanne',
        'resultLocation': u'X=152884.58&Y=538433.38&zoom=10'
    },
    {
        'searchText': u'basel',
        'resultTitle': u'Basel (BS)',
        'resultLocation': 'X=267108.81&Y=611722.92&zoom=6'
    }
]

swissSearchParamTests = [
    {
        'searchText': u'chemin des caves 11',
        'oneResOnly': True,
        'resultTitle': u'Chemin des Caves 11 1040 Echallens',
        'resultLocation': u'X=166095.47&Y=538398.31&zoom=10'
    },
    {
        'searchText': u'oliviers vinzel 1',
        'oneResOnly': False,
        'resultTitle': u'Chemin des Oliviers 8 1184 Vinzel',
        'resultLocation': u'X=144686.36&Y=511163.53&zoom=10'
    }
]


def runSwissSearchTest(driver, target, is_top_browser):
    print 'Start Swissearch tests'

    # swissearch parameter with multiple results
    targetUrl = target + '/?lang=de'
    driver.get(targetUrl)
    # wait until the page is loaded. We know this when the title contain
    # (S)chweiz
    try:
        WebDriverWait(driver, 10).until(EC.title_contains('chweiz'))
    except Exception as e:
        print '-----------'
        print str(e)
        raise Exception('Unable to load map.geo.admin page!')

    # Test interactions with input element, dropdown list, clear btn etc..
    inputSearchEl = driver.find_element_by_class_name('ga-search-input')
    for searchTest in searchLocationTests:
        testSearchInput(driver, inputSearchEl, searchTest)

    # Test swissearch parameter in permalink
    for swissSearchTest in swissSearchParamTests:
        testSwissSearchParameter(driver, target, inputSearchEl, swissSearchTest)
